File: benchmarks-k8s-vs-k3s/benchmarks/remote-write.py
import json
import os
import glob
import argparse
from datetime import datetime
import requests
import snappy
from prometheus_client.samples import Sample
from prometheus_client.exposition import generate_latest
from prometheus_remote_write import RemoteWriteClient, TimeSeries as RWTimeSeries

# Alternative implementation using just requests + protobuf
import struct

def parse_timestamp(ts_str):
    """Parse ISO 8601 timestamp to milliseconds since epoch"""
    dt = datetime.fromisoformat(ts_str.replace('Z', '+00:00'))
    return int(dt.timestamp() * 1000)

def create_protobuf_payload(metrics_data, instance):
    """
    Create Prometheus remote write protobuf payload manually
    This is a simplified version - for production use prometheus-client or similar
    """
    from google.protobuf import timestamp_pb2
    from prometheus_client import core
    
    # We'll use a simpler approach with the prometheus_remote_write library
    timeseries_list = []
    
    for metric in metrics_data:
        labels = {
            "__name__": metric["metricName"],
            "instance": instance,
            "job": metric.get("jobName", "density-test")
        }
        
        # Add custom labels if present
        if metric.get("labels"):
            labels.update(metric["labels"])
        
        # Parse timestamp
        timestamp_ms = parse_timestamp(metric["timestamp"])
        
        # Create time series
        ts = {
            "labels": [{"name": k, "value": v} for k, v in labels.items()],
            "samples": [{
                "value": float(metric["value"]),
                "timestamp": timestamp_ms
            }]
        }
        
        timeseries_list.append(ts)
    
    return timeseries_list

def send_remote_write(url, timeseries_data):
    """Send data via Prometheus remote write protocol"""
    # This is a simplified implementation
    # For production, use: pip install prometheus-remote-write
    
    import remote_write_pb2  # You'll need to generate this from prometheus.proto
    
    # For now, we'll use a simpler JSON-based approach compatible with VictoriaMetrics
    # and some Prometheus setups
    pass

def send_via_importapi(url, metrics_data, instance):
    """
    Send metrics using import API (works with VictoriaMetrics and some Prometheus builds)
    This is simpler and preserves timestamps
    """
    lines = []
    
    for metric in metrics_data:
        labels_dict = {
            "instance": instance,
            "job": metric.get("jobName", "density-test")
        }
        
        # Add custom labels if present
        if metric.get("labels"):
            labels_dict.update(metric["labels"])
        
        # Build label string
        label_str = ",".join([f'{k}="{v}"' for k, v in labels_dict.items()])
        
        # Parse timestamp to milliseconds
        timestamp_ms = parse_timestamp(metric["timestamp"])
        
        # Format: metric_name{labels} value timestamp
        line = f'{metric["metricName"]}{{{label_str}}} {metric["value"]} {timestamp_ms}'
        lines.append(line)
    
    # Send all metrics for this file
    payload = "\n".join(lines)
    
    # VictoriaMetrics import format
    import_url = url.replace("/api/v1/write", "/api/v1/import/prometheus")
    
    response = requests.post(
        import_url,
        data=payload,
        headers={"Content-Type": "text/plain"}
    )
    
    if response.status_code not in [200, 204]:
        raise Exception(f"Failed to send metrics: {response.status_code} - {response.text}")

def process_file(filepath, instance, prometheus_url):
    """Process a single JSON file and send to Prometheus"""
    with open(filepath, 'r') as f:
        metrics_data = json.load(f)
    
    metric_name = os.path.splitext(os.path.basename(filepath))[0]
    print(f"  Processing: {filepath} (metric: {metric_name})")
    
    try:
        send_via_importapi(prometheus_url, metrics_data, instance)
        print(f"  ✓ Pushed {metric_name} for {instance}")
        return True
    except Exception as e:
        print(f"  ✗ Error: {e}")
        return False

def main():
    parser = argparse.ArgumentParser(description='Push kube-burner metrics to Prometheus via remote write')
    parser.add_argument(
        '--url',
        default='http://localhost:8428/api/v1/import/prometheus',
        help='Prometheus/VictoriaMetrics URL (default: VictoriaMetrics import endpoint)'
    )
    parser.add_argument(
        '--prometheus',
        action='store_true',
        help='Use Prometheus remote write endpoint instead of import API'
    )
    
    args = parser.parse_args()
    
    if args.prometheus:
        prometheus_url = args.url.replace('/api/v1/import/prometheus', '/api/v1/write')
        print("Note: Prometheus remote write requires protobuf. Using import API is recommended.")
        print(f"Using URL: {prometheus_url}")
    else:
        prometheus_url = args.url
        print(f"Using import API URL: {prometheus_url}")
    
    # Process k3s and k8s directories
    for instance in ['k3s', 'k8s']:
        kube_burner_dir = os.path.join(instance, 'kube-burner')
        
        if not os.path.exists(kube_burner_dir):
            print(f"Skipping {instance} - kube-burner directory not found")
            continue
        
        print(f"\n=== Processing {instance} ===")
        
        # Find all JSON files
        json_files = glob.glob(os.path.join(kube_burner_dir, '*.json'))
        
        for json_file in json_files:
            process_file(json_file, instance, prometheus_url)
    
    print("\n✓ All metrics pushed successfully!")

if __name__ == '__main__':
    main()

File: benchmarks-k8s-vs-k3s/utils/remote-write/remote-write.go
package main

import (
	"bufio"
	"bytes"
	"context"
	"encoding/json"
	"flag"
	"fmt"
	"io/ioutil"
	"log"
	"net/http"
	"os"
	"os/exec"
	"path/filepath"
	"strings"
	"time"

	"github.com/gogo/protobuf/proto"
	"github.com/golang/snappy"
	"github.com/prometheus/prometheus/prompb"
)

// KubeBurnerMetric represents a single metric from kube-burner JSON
type KubeBurnerMetric struct {
	Timestamp  string            `json:"timestamp"`
	Labels     map[string]string `json:"labels"`
	Value      float64           `json:"value"`
	MetricName string            `json:"metricName"`
	JobName    string            `json:"jobName"`
}

type PortForwardConfig struct {
	Namespace   string
	ServiceName string
	Port        string
}

func main() {

	cfg, err := parseFlags()
	if err != nil {
		log.Fatal(err)
	}

	remoteWriteMetricsURL := fmt.Sprintf("http://localhost:%s/api/v1/write", cfg.Port)

	ctx, cancel := context.WithCancel(context.Background())
	defer cancel()

	// Start port-forward
	pfCmd, err := startServicePortForward(ctx, cfg)
	if err != nil {
		log.Fatalf("failed to start port-forward: %v", err)
	}
	defer func() {
		log.Println("Stopping VictoriaMetrics/Prometheus port-forward")
		cancel()
		_ = pfCmd.Wait()
	}()

	// Get absolute path -> Binary will succeed regardless of
	exePath, err := os.Executable()
	if err != nil {
		log.Fatal(err)
	}

	exeDir := filepath.Dir(exePath)

	// utils/golang/build -> go up 3 times
	repoRoot := filepath.Join(exeDir, "..", "..", "..")
	repoRoot, _ = filepath.Abs(repoRoot)

	// Process k3s and k8s directories
	for _, instance := range []string{"k3s", "k8s"} {
		kubeBurnerDir := filepath.Join(repoRoot, "benchmarks", instance, "kube-burner")

		if _, err := os.Stat(kubeBurnerDir); os.IsNotExist(err) {
			log.Printf("Skipping %s - directory not found\n", kubeBurnerDir)
			continue
		}

		log.Printf("=== Processing %s ===\n", instance)

		// Find all JSON files
		files, err := filepath.Glob(filepath.Join(kubeBurnerDir, "*.json"))
		if err != nil {
			log.Fatalf("Error finding JSON files: %v", err)
		}

		for _, file := range files {
			metricName := filepath.Base(file[:len(file)-5]) // Remove .json extension
			log.Printf("  Processing: %s (metric: %s)\n", file, metricName)

			if err := processFile(file, instance, remoteWriteMetricsURL); err != nil {
				log.Printf("  ✗ Error processing %s: %v\n", file, err)
			} else {
				log.Printf("  ✓ Pushed %s for %s\n", metricName, instance)
			}
		}
		fmt.Println()
	}

	log.Println("All metrics pushed successfully!")
}

func parseFlags() (*PortForwardConfig, error) {
	var (
		namespace   = flag.String("namespace", "monitoring", "Kubernetes namespace")
		serviceName = flag.String("service-name", "", "Service name (required)")
		remoteWrite = flag.String("remote-write", "", "Remote write target: prometheus | victoria-metrics (required)")
	)

	flag.Parse()

	if *serviceName == "" {
		log.Fatal("--service-name is required")
	}

	if *remoteWrite == "" {
		log.Fatal("--remote-write is required (prometheus | victoria-metrics)")
	}

	var port string

	switch *remoteWrite {
	case "prometheus":
		port = "9090"
	case "victoria-metrics":
		port = "8428"
	default:
		log.Fatalf("invalid --remote-write value: %s (allowed: prometheus | victoria-metrics)", *remoteWrite)
	}

	return &PortForwardConfig{
		Namespace:   *namespace,
		ServiceName: *serviceName,
		Port:        port,
	}, nil

}

func startServicePortForward(ctx context.Context, cfg *PortForwardConfig) (*exec.Cmd, error) {
	cmd := exec.CommandContext(
		ctx,
		"kubectl",
		"port-forward",
		"-n", cfg.Namespace,
		fmt.Sprintf("svc/%s", cfg.ServiceName),
		fmt.Sprintf("%s:%s", cfg.Port, cfg.Port),
	)

	stdout, err := cmd.StdoutPipe()
	if err != nil {
		return nil, err
	}
	cmd.Stderr = os.Stderr

	if err := cmd.Start(); err != nil {
		return nil, err
	}

	// Wait until port-forward is ready
	scanner := bufio.NewScanner(stdout)
	for scanner.Scan() {
		line := scanner.Text()
		log.Println("[kubectl]", line)
		if strings.Contains(line, "Forwarding from") {
			log.Println("Prometheus port-forward ready")
			return cmd, nil
		}
	}

	return nil, fmt.Errorf(`port-forward command could not be completed. Please verify that:
    * The specified port is not already in use
    * The service name is correct and exists in the specified namespace`)
}

func processFile(filename, instance, prometheusURL string) error {
	// Read JSON file
	data, err := ioutil.ReadFile(filename)
	if err != nil {
		return fmt.Errorf("reading file: %w", err)
	}

	// Parse JSON
	var metrics []KubeBurnerMetric
	if err := json.Unmarshal(data, &metrics); err != nil {
		return fmt.Errorf("parsing JSON: %w", err)
	}

	// Convert to Prometheus TimeSeries
	var timeSeries []prompb.TimeSeries

	for _, metric := range metrics {
		// Parse timestamp
		t, err := time.Parse(time.RFC3339Nano, metric.Timestamp)
		if err != nil {
			return fmt.Errorf("parsing timestamp: %w", err)
		}
		timestampMs := t.UnixNano() / int64(time.Millisecond)

		// Build labels
		labels := []prompb.Label{
			{Name: "__name__", Value: metric.MetricName},
			{Name: "instance", Value: instance},
			{Name: "job", Value: metric.JobName},
		}

		// Add custom labels if present
		for k, v := range metric.Labels {
			labels = append(labels, prompb.Label{Name: k, Value: v})
		}

		// Create TimeSeries
		ts := prompb.TimeSeries{
			Labels: labels,
			Samples: []prompb.Sample{
				{
					Value:     metric.Value,
					Timestamp: timestampMs,
				},
			},
		}

		timeSeries = append(timeSeries, ts)
	}

	// Create WriteRequest
	writeRequest := &prompb.WriteRequest{
		Timeseries: timeSeries,
	}

	// Marshal to protobuf
	data, err = proto.Marshal(writeRequest)
	if err != nil {
		return fmt.Errorf("marshaling protobuf: %w", err)
	}

	// Compress with snappy
	compressed := snappy.Encode(nil, data)

	// Send HTTP request
	req, err := http.NewRequest("POST", prometheusURL, bytes.NewReader(compressed))
	if err != nil {
		return fmt.Errorf("creating request: %w", err)
	}

	req.Header.Set("Content-Encoding", "snappy")
	req.Header.Set("Content-Type", "application/x-protobuf")
	req.Header.Set("X-Prometheus-Remote-Write-Version", "0.1.0")

	client := &http.Client{Timeout: 10 * time.Second}
	resp, err := client.Do(req)
	if err != nil {
		return fmt.Errorf("sending request: %w", err)
	}
	defer resp.Body.Close()

	if resp.StatusCode/100 != 2 {
		body, _ := ioutil.ReadAll(resp.Body)
		return fmt.Errorf("bad status: %d, body: %s", resp.StatusCode, string(body))
	}

	return nil
}
